gobblin-metrics-libs/gobblin-metrics-base/src/main/java/org/apache/gobblin/metrics/reporter/FileFailureEventReporter.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.metrics.reporter;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintStream;
+
+import org.apache.gobblin.metrics.GobblinTrackingEvent;
+import org.apache.gobblin.metrics.MetricContext;
+import org.apache.gobblin.metrics.event.FailureEvent;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import com.google.common.base.Charsets;
+
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * An {@link OutputStreamEventReporter} reports only {@link FailureEvent}. It won't create
+ * the failure log file until a failure event is processed
+ */
+@Slf4j
+public class FileFailureEventReporter extends OutputStreamEventReporter {
+  private final FileSystem fs;
+  private final Path failureLogFile;
+  private boolean hasSetupOutputStream;
+
+  public FileFailureEventReporter(MetricContext context, FileSystem fs, Path failureLogFile)
+      throws IOException {
+    super(OutputStreamEventReporter.forContext(context));
+    this.fs = fs;
+    this.failureLogFile = failureLogFile;
+    hasSetupOutputStream = false;
+  }
+
+  @Override
+  public void addEventToReportingQueue(GobblinTrackingEvent event) {
+    if (FailureEvent.isFailureEvent(event)) {
+      setupOutputStream();
```

Comment:
What's the purpose of lazy initialization?

Reply:
The goal of the failure logs is to quickly find the cause of job failure. Lazy initialization prevents creating empty failure log files when there is no error in a Task, which reduces the distraction for finding the cause.